nft-client/src/cts1.rs

```diff
@@ -0,0 +1,377 @@
+        }
+        let url = String::from_utf8(bytes)?; //.map_err(|e| ParseError::default())?
+        let hash = Option::<Sha256>::deserial(source)?;
+        Ok(MetadataUrl { url, hash })
+    }
+}
+
+/// Smart contract logged event, part of the CTS1 specification.
+pub enum Event {
+    /// Transfer of an amount of tokens
+    Transfer {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        from:     Address,
+        to:       Address,
+    },
+    /// Minting an amount of tokens
+    Mint {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Burning an amount of tokens
+    Burn {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Add/Remove an address as operator for some other address.
+    UpdateOperator {
+        update:   OperatorUpdate,
+        owner:    Address,
+        operator: Address,
+    },
+    /// Provide an URL with the metadata for a certain token.
+    TokenMetadata {
+        token_id:     TokenIdVec,
+        metadata_url: MetadataUrl,
+    },
+    /// Custom event outside of the CTS1 specification.
+    Unknown,
+}
+
+impl Display for Event {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        match self {
+            Event::Transfer {
+                token_id,
+                from,
+                to,
+                amount,
+            } => {
+                if *amount > 0 {
+                    write!(
+                        f,
+                        "Transferred token with ID {} from {} to {}",
+                        token_id,
+                        address_display(from),
+                        address_display(to)
+                    )?;
+                }
+            }
+            Event::Mint {
+                token_id,
+                amount,
+                owner,
+            } => {
+                if *amount > 0 {
+                    write!(
+                        f,
+                        "Minted token with ID {} for {}",
+                        token_id,
+                        address_display(owner)
+                    )?;
+                }
+            }
+            Event::Burn {
+                token_id,
+                amount,
+                owner,
+            } => {
+                if *amount > 0 {
+                    write!(
+                        f,
+                        "Burned token with ID {} for {}",
+                        token_id,
+                        address_display(owner)
+                    )?;
+                }
+            }
+            Event::UpdateOperator {
+                update,
+                owner,
+                operator,
+            } => {
+                let operation = match update {
+                    OperatorUpdate::Remove => "Remove",
+                    OperatorUpdate::Add => "Add",
+                };
```

Comment:
Instead of this, the correct thing to do would be to add a `Display` implementation for `OperatorUpdate`.